### PR TITLE
Hidden addkeytable command

### PR DESCRIPTION
**Pull request by jigold**
- Takes a named key(s) expression and a named aggregator expression and creates a new table with columns for each of the keys and columns for the results of each aggregator given the eval context for those keys.
- Currently writes out a text file

### CODE REVIEW

File: src/main/scala/org/broadinstitute/hail/driver/AddKeyTable.scala [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+      for (i <- array.indices) {
+        array(i).seqOp(aggT)
+      }
+      array
+    }
+
+    vds.mapPartitionsWithAll { it =>
+      it.map { case (v, va, s, sa, g) =>
+        ec.setAll(v, va, s, sa, g)
+        val key = keyF().toIndexedSeq
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/AddKeyTable.scala]: Heads up: we've learned that array.toIndexedSeq produces a Vector, which is an absolutely horrible data structure.  casting an array to an indexed seq with `arr: IndexedSeq[T]` produces a WrappedArray, which is much better.

File: src/main/scala/org/broadinstitute/hail/driver/AddKeyTable.scala [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    ec.set(4, vds.globalAnnotation)
+    aggregationEC.set(4, vds.globalAnnotation)
+
+    val (keyNames, keyParseTypes, keyF) = Parser.parseNamedArgs(keyCond, ec)
+    val (aggNames, aggParseTypes, aggF) = Parser.parseNamedArgs(aggCond, ec)
+
+    if (keyNames.isEmpty)
+      fatal("this module requires one or more named expr arguments as keys")
+    if (aggNames.isEmpty)
+      fatal("this module requires one or more named expr arguments to aggregate by key")
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/AddKeyTable.scala]: Are these strictly necessary?  Empty keys means the key is an empty struct and that allows you to do global aggregation.  (A scalar is the same as a single-row table.)  Empty aggregator is less obviously useful, but maybe you are just interested in the set of aggregation keys that appear...

File: src/main/scala/org/broadinstitute/hail/driver/AddKeyTable.scala [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    val (zVals, _, combOp, resultOp) = Aggregators.makeFunctions(aggregationEC)
+    val zvf = () => zVals.indices.map(zVals).toArray
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/AddKeyTable.scala]: Is this necessary?  This just copies the zVals array (but not the underlying aggregators.)  aggregateByKey should copy the zero value with serialization (as our aggregateBy... VSM functions do.)

Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/driver/AddKeyTable.scala]: Ah, I see, you copied this from AggregateIntervals.  I think this shouldn't be necessary there, either.

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/AddKeyTable.scala]: Yep, Aggregators are a jungle.  I take responsibility

**Comment by cseed** [comment]
Superseded by https://github.com/hail-is/hail/pull/1089

**Pull request closed without merging**